### ISSUE TITLE
sincos unit test fails when compiled with clang-8.0

### ISSUE THREAD

**Issue by mkatliar** [stack-of-tasks/pinocchio#796]
Build commands:
```bash
cd build
cmake .. -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_BUILD_TYPE=Release
make sincos
```

Run test:
```bash
unittest/sincos
```

Test output:
```
Running 1 test case...
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(33): [1;31;49merror: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed[0;39;49m
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(33): [1;31;49merror: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed[0;39;49m
...
"sincosTest/test_sincos": check sin_value == sin_value_ref has failed[0;39;49m
*** 1287 failures are detected in the test module "sincosTest
```

By changing `BOOST_CHECK` lines in the code to
```c++
    BOOST_CHECK_EQUAL(sin_value, sin_value_ref);
    BOOST_CHECK_EQUAL(cos_value, cos_value_ref);
```
I can see the values:
```
...
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(25): error: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed [0.33331491171648064 != 0.3333149117164807]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(25): error: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed [0.81408624132169871 != 0.8140862413216986]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(25): error: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed [0.24940469018008196 != 0.24940469018008193]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(26): error: in "sincosTest/test_sincos": check cos_value == cos_value_ref has failed [0.95572248566259776 != 0.95572248566259765]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(26): error: in "sincosTest/test_sincos": check cos_value == cos_value_ref has failed [0.86869987885449307 != 0.86869987885449296]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(25): error: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed [0.5078055634522779 != 0.50780556345227801]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(26): error: in "sincosTest/test_sincos": check cos_value == cos_value_ref has failed [0.94032322684914449 != 0.9403232268491446]
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(25): error: in "sincosTest/test_sincos": check sin_value == sin_value_ref has failed [0.36897122731672255 != 0.36897122731672249]
```
So the difference between the actual and the reference sin and cos values is very small but not 0.

Compiler version:
```
clang version 8.0.0-3 (tags/RELEASE_800/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Exact compile command:
```bash
/usr/bin/clang++   -pedantic -Wno-long-long -Wall -Wextra -Wcast-align -Wcast-qual -Wformat -Wwrite-strings -Wconversion  -O3 -DNDEBUG  -rdynamic CMakeFiles/sincos.dir/sincos.cpp.o  -o sincos -Wl,-rpath,/home/kotlyar/software/pinocchio/build/src ../src/libpinocchio.so -lboost_unit_test_framework -lboost_filesystem -lboost_system -lboost_serialization -lurdfdom_sensor -lurdfdom_model_state -lurdfdom_model -lurdfdom_world -lconsole_bridge
```

CPU info: [cpuinfo.txt](https://github.com/stack-of-tasks/pinocchio/files/3184741/cpuinfo.txt)

With g++ as the compiler, the test passes.

**Comment by mkatliar** [comment]
In Debug configuration, I get only 1 failure in the test:
```
Running 1 test case...
alpha = 0.7053, cos_value = 0.761417, cos_value_ref = 0.761417, delta = 1.11022e-16
/home/kotlyar/software/pinocchio/unittest/sincos.cpp(34): error: in "sincosTest/test_sincos": check cos_value == cos_value_ref has failed

*** 1 failure is detected in the test module "sincosTest"
```
Compiler command line:
```bash
/usr/bin/clang++   -pedantic -Wno-long-long -Wall -Wextra -Wcast-align -Wcast-qual -Wformat -Wwrite-strings -Wconversion  -g  -rdynamic CMakeFiles/sincos.dir/sincos.cpp.o  -o sincos -Wl,-rpath,/home/kotlyar/software/pinocchio/build/src ../src/libpinocchio.so -lboost_unit_test_framework -lboost_filesystem -lboost_system -lboost_serialization -lurdfdom_sensor -lurdfdom_model_state -lurdfdom_model -lurdfdom_world -lconsole_bridge
```

**Comment by mkatliar** [comment]
The issue appears only for Scalar = double. For other types, the match is exact.

**Comment by mkatliar** [comment]
Proposed solution: for the `double` type, check that the values are close with ~1e-16 absolute tolerance, instead of exact match.

**Comment by stonneau** [comment]
A more general remark is that as far as I know 
BOOST_CHECK_CLOSE is not recommended to compare to 0, but that BOOST_CHECK_SMALL should rather be used. 
See
https://stackoverflow.com/questions/1093453/difference-between-boost-check-close-and-boost-check-close-fraction#1093530

and https://www.boost.org/doc/libs/1_34_0/libs/test/doc/components/test_tools/floating_point_comparison.html

**Comment by mkatliar** [comment]
@stonneau we don't compare to 0, we compare two numbers -- the actual and the reference sin/cos value. What I do is
```c++
BOOST_CHECK_CLOSE_FRACTION(sin_value, sin_value_ref, sinCosTolerance<Scalar>());
BOOST_CHECK_CLOSE_FRACTION(cos_value, cos_value_ref, sinCosTolerance<Scalar>());
```

**Comment by jmirabel** [comment]
To be sure you never compare to zero, you could replace
```c++
Scalar alpha = (Scalar)std::rand()/(Scalar)RAND_MAX;
```
by
```c++
Scalar alpha = (Scalar)0.1 + (Scalar)std::rand()/(Scalar)RAND_MAX;
```
This way we are sure that both the sine and the cosine are not close to zero.

**Comment by jcarpent** [comment]
I've suggested a workaround proposed by Boost.Test itself in #797, which does not require such a trick.

**Comment by jcarpent** [comment]
Solved by #797.
